From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ophuscado <postbox@ophuscado.com>
Date: Sun, 11 Oct 2026 14:08:14 +0000
Subject: [PATCH 3/9] DEBUG: thunderbolt: apple: Add cable_info_or parameter

Run-time override for the cable information word given to the host
router firmware, used to confirm the 20 Gb/s flag fix before it was
written. Kept as its own commit so it can be dropped.
---
 drivers/thunderbolt/apple.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/thunderbolt/apple.c b/drivers/thunderbolt/apple.c
index 173a78c7aa..94f3d79602 100644
--- a/drivers/thunderbolt/apple.c
+++ b/drivers/thunderbolt/apple.c
@@ -140,6 +140,11 @@
 #define TB_VSE_CAP_APPLE_CABLE_INFO_LEGACY_ADAPTER BIT(9)
 #define TB_VSE_CAP_APPLE_CABLE_INFO_TBT2_3 BIT(10)
 
+static unsigned int cable_info_or;
+module_param(cable_info_or, uint, 0644);
+MODULE_PARM_DESC(cable_info_or,
+		 "Debug: bits OR-ed into the cable information word of a present cable");
+
 /**
  * struct apple_cio - Apple Converged I/O host router complex
  * @dev: ACIO device
@@ -856,6 +861,14 @@ static int apple_cio_tbt_switch_set(struct typec_thunderbolt_switch_dev *sw,
 		break;
 	}
 
+	/* Debug aid: lets cable-information bits be tested without rebuilding. */
+	if (acio->target_cable_info && cable_info_or) {
+		dev_info(acio->dev, "cable info 0x%x overridden to 0x%x\n",
+			 acio->target_cable_info,
+			 acio->target_cable_info | cable_info_or);
+		acio->target_cable_info |= cable_info_or;
+	}
+
 	if (acio->target_cable_info == acio->current_cable_info)
 		return 0;
 
